src/layers/generic.cpp

```diff
@@ -439,18 +442,26 @@ namespace marian {
     std::string name = opt<std::string>("prefix");
     int dimVoc = opt<int>("dimVocab");
     int dimEmb = opt<int>("dimEmb");
+    int dimFactorEmb = opt<int>("dimFactorEmb");
 
     bool fixed = opt<bool>("fixed", false);
 
+    // Embedding layer initialization should depend only on embedding size, hence fanIn=false
+    auto initFunc = inits::glorotUniform(/*fanIn=*/false, /*fanOut=*/true); // -> embedding vectors have roughly unit length
+
     factoredVocab_ = FactoredVocab::tryCreateAndLoad(options_->get<std::string>("vocab", ""));
     if (factoredVocab_) {
       dimVoc = (int)factoredVocab_->factorVocabSize();
       LOG_ONCE(info, "[embedding] Factored embeddings enabled");
+      if (opt<std::string>("factorsCombine") == "concat") {
+        ABORT_IF(dimFactorEmb == 0, "Embedding: If concatenation is chosen to combine the factor embeddings, a factor embedding size should be specified.");
+        int numberOfFactors = (int) factoredVocab_->getTotalFactorCount();
+        dimVoc -= numberOfFactors;
+        FactorEmbMatrix_ = graph_->param("factor_" + name, {numberOfFactors, dimFactorEmb}, initFunc, fixed);
+        LOG_ONCE(info, "[embedding] Combining factors concatenation enabled");
+      }      
```

Comment:
If concatenation is chosen the factors will be embedded with a different matrix from the lemma embeddings.